cpp/include/raft/neighbors/detail/cagra/cagra_build.cuh

```diff
@@ -264,4 +264,82 @@ void build_knn_graph(raft::resources const& res,
   graph::sort_knn_graph(res, dataset, knn_graph_internal);
 }
 
+template <typename IdxT = uint32_t,
+          typename g_accessor =
+            host_device_accessor<std::experimental::default_accessor<IdxT>, memory_type::host>>
+void optimize(raft::resources const& res,
+              mdspan<IdxT, matrix_extent<int64_t>, row_major, g_accessor> knn_graph,
+              raft::host_matrix_view<IdxT, int64_t, row_major> new_graph)
+{
+  using internal_IdxT = typename std::make_unsigned<IdxT>::type;
+
+  auto new_graph_internal = raft::make_host_matrix_view<internal_IdxT, int64_t>(
+    reinterpret_cast<internal_IdxT*>(new_graph.data_handle()),
+    new_graph.extent(0),
+    new_graph.extent(1));
+
+  using g_accessor_internal =
+    host_device_accessor<std::experimental::default_accessor<internal_IdxT>, memory_type::host>;
+  auto knn_graph_internal =
+    mdspan<internal_IdxT, matrix_extent<int64_t>, row_major, g_accessor_internal>(
+      reinterpret_cast<internal_IdxT*>(knn_graph.data_handle()),
+      knn_graph.extent(0),
+      knn_graph.extent(1));
+
+  cagra::detail::graph::optimize(res, knn_graph_internal, new_graph_internal);
+}
+
+template <typename T,
+          typename IdxT = uint32_t,
+          typename Accessor =
+            host_device_accessor<std::experimental::default_accessor<T>, memory_type::host>>
+index<T, IdxT> build(raft::resources const& res,
+                     const index_params& params,
+                     mdspan<const T, matrix_extent<int64_t>, row_major, Accessor> dataset,
+                     std::optional<float> refine_rate                   = std::nullopt,
+                     std::optional<ivf_pq::index_params> build_params   = std::nullopt,
+                     std::optional<ivf_pq::search_params> search_params = std::nullopt)
```

Comment:
Allowing these parameters is the difference between the public and the detail:: version of build.

Reply:
This is a bit conflicting because you could also build `NN_DESCENT` with this method. Could we allow the base class `ann::index_params` as the function argument here, and then cast it to either IVF_PQ or NN_DESCENT index_params based on `cagra::index_params::graph_build_algo`?

Reply:
That would require us to change `ann::index_params` into a polymorphic class. If we change that, then [this assertion would fail](https://github.com/rapidsai/raft/blob/branch-23.12/cpp/include/raft/neighbors/cagra_types.hpp#L120) (and also similar assertions for ivf_flat and ivf_pq).

I do not know what was the motivation behind keeping the index params as aggregate type. I would prefer to return to such question in a separate issue, and keep this PR a non-breaking change.

I have added an `optional<nn_descent_params>`, so that we can fine set these through the config file.